### ISSUE TITLE
How to use rust-analyzer as a crate?

### ISSUE THREAD

**Issue by lcswillems** [rust-lang/rust-analyzer#14466]
Hi,

I see that rust-analyzer is published as a crate: https://crates.io/crates/ra_ap_rust-analyzer

Is there any example on how to use it?

**Comment by lnicola** [comment]
Two of them, even! https://crates.io/crates/ra_ap_rust-analyzer/reverse_dependencies

**Comment by oooutlk** [comment]
I have the same issue. Just created a [repository](https://github.com/oooutlk/try_rust_analyzer) for demonstration.

I did not dig deep into this crate, just tried to make the compiler happy.

Fortunately, I did it and got the right result.
Unfortunately, the running speed is way too slow.

Need some help. Wish to have "hello, world" examples.

**Comment by flodiebold** [comment]
Right now providing a library is not our goal, the crates are provided "as-is" without any guarantees and we have no concerns about breaking their APIs. So we're unlikely to spend time providing examples for them.

That said,

> Unfortunately, the running speed is way too slow.

What kind of performance are you getting and did you expect? Keep in mind that analyzing a crate in order to e.g. provide definitions involves 1) building all build-time dependencies like proc macros, and 2) fully macro-expanding and name-resolving the crate and its dependencies on the item level (not in function bodies, but everything outside).

**Comment by oooutlk** [comment]
> Right now providing a library is not our goal, the crates are provided "as-is" without any guarantees and we have no concerns about breaking their APIs.

So the stable way is manipulating JSON-RPCs to rust analyzer binaries.

> What kind of performance are you getting and did you expect? 

`cargo run —debug` spent 100-120 seconds in locating the definition of [`do_nothing()` ](https://github.com/oooutlk/try_rust_analyzer/blob/main/src/main.rs#L47) which is unacceptable.

`cargo run —release` finished in 6-8 seconds, much better. And the second `go to definition` request  completed immediately. I guess 6 seconds are the “boot time” of the analysis engine?  That’s OK for me.

**Comment by lnicola** [comment]
Yeah, most of that time is analyzing the sysroot (standard library). If you can keep your program running, it shouldn't be a problem. The analysis is cached, so follow-up commands should be fast, but you can also disable cache priming (enabled by default) to trade a bit of extra latency on some requests for faster / more efficient start-up.

**Comment by Qi-Zhan** [comment]
> Do we do cache priming when you use RA as a library? 🤔

maybe he wants to do some code analysis tasks like me. I want

**Comment by ChayimFriedman2** [comment]
Things have changed since and this is probably irrelevant.
